### DISCUSSION TITLE
Cloudflare Workers not work with Countries

### DISCUSSION THREAD

**Discussion by viriatusX** [plausible/analytics#3978] [Self-Hosted Support] [answered]
### Past Issues Searched

- [X] I have searched open and closed issues to make sure that the bug has not yet been reported

### Issue is a Bug Report

- [X] This is a bug report and not a feature request, nor asking for self-hosted support

### Using official Plausible Cloud hosting or self-hosting?

Self-hosting

### Describe the bug

When you enable the Plausible method with Cloudflare workers (https://plausible.io/docs/proxy/guides/cloudflare).
It does not update the visitors' country map in the UI.

### Expected behavior

Plausible should update the UI, the visitor map by country, by having CloudFlare workers enabled.

**Comment by ruslandoga** [comment]
👋 @viriatusX 

The last time I tried cf workers, everything was working as expected: https://github.com/plausible/analytics/discussions/2137#discussioncomment-6349831

Could you please share more of your setup?

**Comment by viriatusX** [comment]
Hi @ruslandoga 

All statistics are displayed correctly, except the Map with the country / region / city of the visitors. It does not show anything.
When I deactivate the Worker, everything works perfectly.

I have Plausible with Nginx Proxy Manager (without additional HTTP headers.
I have the Plausible domain as "Proxied".
I have tried with Plausible remote:
`iex> Plausible.Geo.lookup("1.1.1.1.1")`
And it shows the country and everything ok.

The problem is that it is not reflected in the UI.

**Comment by ruslandoga** [comment]
Would you be able to share your configuration files, including nginx, worker scripts, and anything else that seems relevant?

**Comment by ruslandoga** [comment]
You can also trace the incoming request to see what headers are included with recon_trace. There have been several discussions where it helped resolve IP geo location issues. https://github.com/plausible/analytics/discussions/3248#discussioncomment-6670217 https://github.com/plausible/analytics/discussions/2767#discussioncomment-5367417 etc. (you can search for others using parts of that code snippet)

**Comment by viriatusX** [comment]
I have executed this inside Plausible remote. But I don't know how to check it with my domain. Sorry.
```
iex> headers = fn {:trace, _pid, :call, {_mod, _fun, [%Plug.Conn{req_headers: headers}]}} -> inspect(headers) end
iex> mod_fun_args = {PlausibleWeb.RemoteIp, :get, 1}
iex> how_many_times = 5
iex> :recon_trace.calls(mod_fun_args, how_many_times, formatter: headers)
```


My Worker Script is the same of yours (changing only my domain).

My Nginx conf for the Plausible domain:

```
# ------------------------------------------------------------
# stats.xxxxx.com
# ------------------------------------------------------------


server {
  set $forward_scheme http;
  set $server         "192.168.10.165";
  set $port           3998;

  listen 80;
listen [::]:80;

listen 443 ssl http2;
listen [::]:443 ssl http2;


  server_name xxxxx.com;


  # Let's Encrypt SSL
  include conf.d/include/letsencrypt-acme-challenge.conf;
  include conf.d/include/ssl-ciphers.conf;
  ssl_certificate /etc/letsencrypt/live/npm-23/fullchain.pem;
  ssl_certificate_key /etc/letsencrypt/live/npm-23/privkey.pem;


proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection $http_connection;
proxy_http_version 1.1;

  access_log /data/logs/proxy-host-15_access.log proxy;
  error_log /data/logs/proxy-host-15_error.log warn;


  location / {

    
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $http_connection;
    proxy_http_version 1.1;
    

    # Proxy!
    include conf.d/include/proxy.conf;
  }


  # Custom
  include /data/nginx/custom/server_proxy[.]conf;
}
```

**Comment by viriatusX** [comment]
I execute Plausible GeoIP command with this IP, but in the UI not appear... with Workers enabled.

<img width="556" alt="Captura de pantalla 2024-04-04 a las 23 18 26" src="https://github.com/plausible/analytics/assets/64467637/29e19e5a-7181-4087-82bd-c7247f5c29fd">

<img width="546" alt="Captura de pantalla 2024-04-04 a las 23 18 34" src="https://github.com/plausible/analytics/assets/64467637/ea4b5c69-f53a-494a-b930-84d8139db40c">

**Comment by viriatusX** [comment]
Ant the output for this command (in case it helps.)

<img width="559" alt="Captura de pantalla 2024-04-04 a las 23 20 27" src="https://github.com/plausible/analytics/assets/64467637/6a91581a-7eef-4d8b-8887-12ff7133ee58">

**Comment by ruslandoga** [comment]
> I have executed this inside Plausible remote. But I don't know how to check it with my domain. Sorry.

You would need to visit your site which would make an `/api/event` call. The traces would be printed in the console. Ideally, you'd run it both when the Worker is activated and when it's not.

**Comment by ruslandoga** [comment]
> All statistics are displayed correctly, except the Map with the country / region / city of the visitors. It does not show anything.
> When I deactivate the Worker, everything works perfectly.

I've re-read this now, and it's not clear what exactly is wrong, is the map empty or not displayed at all? Can you show screenshot for when the Worker is activated and when it's not?

**Comment by viriatusX** [comment]
If I access with an IP from Portugal:

**With Worker**

<img width="532" alt="Captura de pantalla 20"
 src="https://github.com/plausible/analytics/assets/64467637/bab13a7e-7014-48b5-9d06-4fdd6f0d6820">

**Without Worker**

<img width="534" alt="Captura de pantalla 2024-04-05 a las 11 37 17" src="https://github.com/plausible/analytics/assets/64467637/ce95ab7c-8cc0-4aec-ae81-796b07a0a3cc">

The map is not updated. The rest of the statistics work OK. If it was a /api/event problem, I think it would not show any statistics.

**Comment by ruslandoga** [comment]
Great, then https://github.com/plausible/analytics/issues/3976#issuecomment-2039307082 should help resolve it. Most likely, the X-Forwarded-For header or similar is dropped somewhere along the way. You can also trace `Plausible.Geo.lookup` to see what exact IP address is used for the lookup of your Portugal request and what info it gets in return.

**Comment by viriatusX** [comment]
> Great, then [#3976 (comment)](https://github.com/plausible/analytics/issues/3976#issuecomment-2039307082) should help resolve it. Most likely, the X-Forwarded-For header or similar is dropped somewhere along the way. You can also trace `Plausible.Geo.lookup` to see what exact IP address is used for the lookup of your Portugal request and what info it gets in return.

Thanks for your help.
Forgive my clumsiness. I have executed this, but then I don't know how to continue. 

```
iex> headers = fn {:trace, _pid, :call, {_mod, _fun, [%Plug.Conn{req_headers: headers}]}} -> inspect(headers) end
iex> mod_fun_args = {PlausibleWeb.RemoteIp, :get, 1}
iex> how_many_times = 5
iex> :recon_trace.calls(mod_fun_args, how_many_times, formatter: headers)
```

**Comment by ruslandoga** [comment]
>  I have executed this, but then I don't know how to continue.

You would need to visit https://hdsplus.co after you run these commands. The traces would be printed in the console. Ideally, you'd run it both when the Worker is activated and when it's not.

You can check the previous discussions I linked for more info :) E.g. https://github.com/plausible/analytics/discussions/2767#discussioncomment-5367417

**Comment by viriatusX** [comment]
Sorry, on which console do I view the trace? The Plausible console does not show anything new.
1. I execute these commands
2. I access to my Web with Chrome and Safari

<img width="948" alt="Captura de pantalla 2024-04-05 a las 12 04 43" src="https://github.com/plausible/analytics/assets/64467637/c84b9b5e-d3b9-45e6-95c3-146579aabf63">

**Comment by ruslandoga** [comment]
If it doesn't show anything, then there are no requests reaching Plausible. Try running it with the Worker deactivated.

**Comment by viriatusX** [comment]
Neither... The Plausible remote console does not show anything new after executing these commands. 
And without Worker everything works fine. Strange.

**Comment by ruslandoga** [comment]
Ok, I see. I guess you are on v2.1.0-rc.0? The module name changed from `RemoteIp` to `RemoteIP` in https://github.com/plausible/analytics/pull/3761. Please run this command instead `:recon_trace.calls({PlausibleWeb.RemoteIP, :get, 1}, 10)`

**Comment by viriatusX** [comment]
Thanks! now yes!

**Comment by ruslandoga** [comment]
Please post the traces as code or text snippet. It would be easier to debug.

**Comment by viriatusX** [comment]
.

**Comment by ruslandoga** [comment]
`x-forwarded-for` are completely different. Are both of these your own requests or real users? Either way, Plausible is picking [the first IP address](https://github.com/plausible/analytics/blob/a7603c9e49d556b4a11d972da993a48772a250fe/lib/plausible_web/remote_ip.ex#L56) in the `x-forwarded-for` list, which is different from `x-real-ip` (whatever it is) which is the last element in the `x-forwarded-for` list. According to https://en.wikipedia.org/wiki/X-Forwarded-For, the client IP address is the first element in the list, so Plausible is correct, at least according to Wikipedia.

**Comment by viriatusX** [comment]
Yes.
I tried again

**Comment by ruslandoga** [comment]
Do you know where `x-forwarded-for` and `x-real-ip` come from? One of them seems to be wrong. According to https://developers.cloudflare.com/fundamentals/reference/http-request-headers/#x-forwarded-for CloudFlare adds the client IP address (the real IP address) to the beginning of the list. And since Plausible is reading it from the beginning of the list, it's not a bug in Plausible.

**Comment by ruslandoga** [comment]
I suggest checking nginx configuration.

**Reply by viriatusX** [reply, marked as answer]
Now yes.... finally...
I added to NGINX:


 ```
 location / {


proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection $http_connection;
proxy_http_version 1.1;
**proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;**
```

And now.. alls works!!

Thanks you so much :)

**Comment by viriatusX** [comment]
**Update: @ruslandoga **

I have already found the "problem".
Plausible works OK and Worker works OK.

But...., if you have the domain as "Proxied" in CloudFlare DNS it does NOT work. Since it already masks the real IP of the Host with a CloudFlare IP.

You can try it if you want.

**Reply by ruslandoga** [reply]
I don't understand, sorry. Does it work or does it not work?

**Reply by ruslandoga** [reply]
And what should I try? :)

**Reply by viriatusX** [reply]
<img width="164" alt="Captura de pantalla 2024-04-05 a las 17 51 46" src="https://github.com/plausible/analytics/assets/64467637/bf8bfb49-d8f0-4a1c-be18-ae01d81d4d96">

If the domain where you have Plausible is "Proxied" the Worker does not work. It updates the statistics except the map (my main problem).

If the domain is as DNS Only, everything works perfectly (Worker included).

I understand, being in Proxied,the Worker can't get the real IP of the user, since the Host is already masquerading with CloudFlare IP due to the Proxied.

**Reply by ruslandoga** [reply]
It should work both ways :) [`cf-connecting-ip`](https://developers.cloudflare.com/fundamentals/reference/http-request-headers/#cf-connecting-ip) is used if available.

**Reply by ruslandoga** [reply]
https://github.com/plausible/analytics/blob/a7603c9e49d556b4a11d972da993a48772a250fe/lib/plausible_web/remote_ip.ex#L8

**Reply by ruslandoga** [reply]
You can follow the same tracing steps as before to find where the problem occurs.

**Reply by viriatusX** [reply]
I seem to have solved it by adding this to NGINX:

```
set_real_ip_from 103.21.244.0/22;
set_real_ip_from 103.22.200.0/22;
set_real_ip_from 103.31.4.0/22;
set_real_ip_from 104.16.0.0/12;
set_real_ip_from 108.162.192.0/18;
set_real_ip_from 131.0.72.0/22;
set_real_ip_from 141.101.64.0/18;
set_real_ip_from 162.158.0.0/15;
set_real_ip_from 172.64.0.0/13;
set_real_ip_from 173.245.48.0/20;
set_real_ip_from 188.114.96.0/20;
set_real_ip_from 190.93.240.0/20;
set_real_ip_from 197.234.240.0/22;
set_real_ip_from 198.41.128.0/17;
set_real_ip_from 2400:cb00::/32;
set_real_ip_from 2405:b500::/32;
set_real_ip_from 2606:4700::/32;
set_real_ip_from 2803:f800::/32;
set_real_ip_from 2c0f:f248::/32;
set_real_ip_from 2a06:98c0::/29;

real_ip_header CF-Connecting-IP;
```
